src/main/java/com/mraof/minestuck/entity/AttackingAnimatedEntity.java

```diff
@@ -0,0 +1,116 @@
+package com.mraof.minestuck.entity;
+
+import net.minecraft.entity.EntityType;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.ai.goal.MeleeAttackGoal;
+import net.minecraft.world.World;
+
+/**
+ * A base class for animated entities with a potentially delayed attack.
+ */
+public abstract class AttackingAnimatedEntity extends AnimatedCreatureEntity
+{
+	/**
+	 * The delay between the start of the animation and the moment the damage lands
+	 */
+	protected int attackDelay;
+	
+	/**
+	 * The delay after an attack and before another start
+	 */
+	protected int attackRecovery;
+	
+	protected AttackingAnimatedEntity(EntityType<? extends AttackingAnimatedEntity> type, World world)
+	{
+		super(type, world);
+	}
+	
+	@Override
+	protected void registerGoals()
+	{
+		super.registerGoals();
+		goalSelector.addGoal(3, new AttackingAnimatedEntity.DelayedAttackGoal(this, 1F, false));
+	}
+	
+	@Override
+	protected void endTimedAction(Actions action)
+	{
+		if(action == Actions.ATTACK)
+		{
+			this.setCurrentAction(Actions.ATTACK_RECOVERY, attackRecovery);
+			performAttack();
+		} else
+			super.endTimedAction(action);
+	}
+	
+	private void performAttack()
+	{
+		this.onAttackEnd();
+		
+		if(getTarget() != null && isInRange(getTarget()))
+		{
+			doHurtTarget(getTarget());
+			// TODO: AOE bounding box collision checks + aoe flag
+		}
+	}
+	
+	private boolean isInRange(LivingEntity target)
+	{
+		double reach = this.getBbWidth() * 2.0F * this.getBbWidth() * 2.0F + target.getBbWidth();
+		return this.distanceToSqr(target) <= reach;
+	}
+	
+	/**
+	 * Starts a long attack against the current target if this entity isn't already performing some action with a duration.
+	 * Useful to sync animations and add extra delays
+	 */
+	private void startAttack()
+	{
+		if(!this.hasTimedAction())
+		{
+			this.setCurrentAction(Actions.ATTACK, attackDelay);
+			
+			this.onAttackStart();
+		}
+	}
+	
+	/**
+	 * Is called when an attack starts. Can be extended to apply effects during an attack.
+	 */
+	protected void onAttackStart()
+	{
+	}
+	
+	/**
+	 * Is called when an attack starts. Can be extended to remove effects once an attack has ended.
+	 */
+	protected void onAttackEnd()
+	{
+	}
```

Comment:
would this be an appropriate location to apply effects or additional modifications to the targetted mob if they are hit? As a random example, if an underling attacks and successfully hits a player and causes the player to catch fire

Reply:
No. This function call happens for an attack end, which does not neccesarily correspond to an attack hit. There eould need to be a different overridable function for that.